docs/index.rst

```diff
@@ -75,11 +75,61 @@ Except the above usage, there are some other use cases:
 
 API Reference
 -------------
-
+:mod:`ubiconfig`
+================
 .. currentmodule:: ubiconfig
+
 .. autofunction:: get_loader
 
 .. autoclass:: UbiConfig
+    :members: load_from_dict
 
 .. autoclass:: Loader
     :members: load, load_all
+
+:mod:`ubiconfig.ubi`
+====================
+
+.. currentmodule:: ubiconfig.ubi
+
+.. autoclass:: GitlabLoader
```

Comment:
Why should this be public API?

Why should anyone directly create an instance of GitlabLoader rather than calling get_loader?

I think exposing these classes as public API simply makes it easier for users of this library to tie themselves to implementation details of how UBI config is stored.  It partially defeats the purpose of abstracting it in the first place.

Reply:
> Why should anyone directly create an instance of GitlabLoader rather than calling get_loader?

Because it's get_loader is meant to be used like this (example from documentation):
>>> loader = get_loader('/my/config/dir)
>>> config = loader.load('path/to/configfile.yaml')
Therefore get_loader returns something which is visible for user. And if it's visible for user and is meant to be used, it should be documented. If I read documentation for get_loader and I can see that if I pass there path, it will return me some LocalLoader and if I pass there url, it returns GitlabLoader. And only thing I know is that it has some load and load_all methods. So basically as user I don't need to know all implementation details or instance members, but I would like to know at least how that loading with the loader works, what is required to be on input, what are possible exceptions that I can expect during loading and similar stuff.
If loaders should meant to be hidden/private, I think it would make more sense to arrange ubiconfig conf loading like this:

>>> conf = ubiconfig.load('/my/config/dir)
>>> conf = ubiconfig.load('https://some.gitlab.host.com/namespace/repo')

Then loaders are completely hidden and there's no need to reference them as some mysterious classes without documentation.

Reply:
> I would like to know at least how that loading with the loader works,

You can always look at the code if you really want to know how something works, but you shouldn't have to design your code in a way which cares about that, and we should aim for an API which discourages building on such details.

> what is required to be on input,

It's already documented on get_loader/Loader.

> what are possible exceptions that I can expect during loading and similar stuff.

If exception handling is really a concern here, we should document on get_loader/Loader what exceptions they're allowed to raise and make sure all implementations match that.  I don't think it's all that important though, in practice does anyone need to do anything with exceptions here which would differ based on the type?  Seems like just "log & re-raise" is what needs to be done by the caller.

> Then loaders are completely hidden and there's no need to reference them as some mysterious classes without documentation.

Loader is documented and the returned values of get_loader implement that interface, so I don't think it's mysterious.  If a dev thinks they need to care *exactly* which class is being returned, they're probably doing something wrong.

You didn't answer "Why should anyone directly create an instance of GitlabLoader rather than calling get_loader".  Let me explain why I think it's a *bad* idea to enable that:

The situation now is that get_loader is the only defined interface for accessing config, and passing different strings to that method is the only way of selecting the config source.  So, everyone using the library needs to stick with just that interface.

That's great for us as it means we can be very confident that any code using ubi-config supports every config source this library itself supports, present and future, and can be switched to another source with trivial or even no code changes needed.

If you expose the specific LocalLoader and GitlabLoader classes, then instead you create uncertainty where any dev can pick and choose from the available methods of loading config, and some will most likely thoughtlessly pick the wrong one.  By that I mean: a dev looks at the docs, thinks "Well, I know the config is stored in gitlab, and I see a GitlabLoader class, guess I'll use that".

Then they've ended up with code which unnecessarily is bound specifically to gitlab, instead of supporting everything this library supports.  The implementation detail that "UBI config is currently stored in a gitlab server" ends up scattered throughout various places instead of consolidated in just this one library.  That's how we'll end up having to do a SPIKE like "UBI config is moving to a different git system, review impact on toolchain" where someone needs to review and design changes to X bits of code rather than updating the one place where UBI config loading was encapsulated.

So, that's the disadvantage of making LocalLoader and GitlabLoader public.  I'm not convinced there is any advantage at all.  Again I ask: Why should anyone directly create an instance of GitlabLoader rather than calling get_loader?

Reply:
> It's already documented on get_loader/Loader.
So if you have lets say 12 loaders, you would rather write 10 pages of what is required for all of the loaders there rather than having simple docs with here's how input argument is resolved and for more info you can look at separate page for every individual loader? That seems to be similar to have super large do_everything method rather then more structured code.
> I don't think it's all that important though, in practice does anyone need to do anything with exceptions here which would differ based on the type? 
What about retries for example?
> You didn't answer "Why should anyone directly create an instance of GitlabLoader rather than calling get_loader".
Because I never said that's what anyone should be doing. Also I wonder what makes you thing that if you reveal existence of those classes anyone will automatically tries to instantiate them.

> If you expose the specific LocalLoader and GitlabLoader classes, then instead you create uncertainty where any dev can pick and choose from the available methods of loading config, and some will most likely thoughtlessly pick the wrong one. By that I mean: a dev looks at the docs, thinks "Well, I know the config is stored in gitlab, and I see a GitlabLoader class, guess I'll use that".

So what? Devel will pick wrong one and then it won't fit for his use case. So he'll read documentation again and see that get_loader should be used instead. It similar to tools that depended on name "productid" in repodata instead of parsing real name from repomd.xml. If you don't follow what's supported, you can get into trouble.

> That's how we'll end up having to do a SPIKE like "UBI config is moving to a different git system
No, we won't end up like that, because we support only ubi-config and ubi-config is only supported library for loading those configurations. If anyone uses anything else (custom forks) or use it incorrectly (not according to documentation or with custom hacks) it's not our responsibility. 

> I'm not convinced there is any advantage at all.
Advantage is that you can have like that documented process for every individual loader that is in this tool separately. And i believe it's also easier to maintain docs in that form. Note that I'm not mentioning doing documentation for other methods than load and load_all.
Or don't have documented Loader at all and have only load method with will do the loading. What is the purpose of having Loader documented anyway (for non-dev users)?

Reply:
> So if you have lets say 12 loaders, you would rather write 10 pages of what is required for all of the loaders there rather than having simple docs with here's how input argument is resolved and for more info you can look at separate page for every individual loader?

No, I rather us document the behavior which needs to be documented, and leave irrelevant details undocumented.

> What about retries for example?

Retries should be implemented within the Loader interface, not outside of it, so everyone using the library will benefit.

> what makes you thing that if you reveal existence of those classes anyone will automatically tries to instantiate them.

If you expose the class with a public-looking name, give it a usable constructor and give it documentation which doesn't warn "Don't create instances of this directly", then most developers looking at that class would think the intent of the library owners is for instances of that class to be directly created.

> So what? Devel will pick wrong one and then it won't fit for his use case. So he'll read documentation again and see that get_loader should be used instead.

This sounds like a description of a dev making a mistake and requiring extra work later to fix it.  I'm proposing that it's better to design the API in a way such that the mistake isn't possible in the first place.

> No, we won't end up like that, because we support only ubi-config and ubi-config is only supported library for loading those configurations. If anyone uses anything else (custom forks) or use it incorrectly (not according to documentation or with custom hacks) it's not our responsibility.

I guess my point wasn't clear.  If you make GitlabLoader class public, then anyone can use supported/documented 
ubi-config API and tie their code into supporting gitlab *only* - rather than supporting *anything ubi-config supports* - requiring extra work later to go and fix it up as new requirements arise.  Creating and using a GitlabLoader seems to be strictly worse than simply calling get_loader.  I don't think we should give users of the library the option to write strictly worse code, or publish docs which could lead them down that path.

> Advantage is that you can have like that documented process for every individual loader that is in this tool separately. And i believe it's also easier to maintain docs in that form.

What Loader-specific docs do you think are missing and need to be added?

If there are docs to be written for e.g. the config file format or the git repo layout, I think those would best not be attached to the Loader classes anyway, since they span several classes - e.g. config file format covers both local and gitlab loaders; git repo layout is not specifically bound to gitlab but is very likely to be the same if we also support other git repos.

> Or don't have documented Loader at all and have only load method with will do the loading. What is the purpose of having Loader documented anyway (for non-dev users)?

I'm not following the point regarding non-dev users, as all users of this library should be developers.

About having a Loader class vs just having a load method, ideally this point would have been raised during initial review of this library.  If you want to change it now, you have to argue not only that dropping Loader class is better, but also that it's worth the cost it takes to change it (update current users of the library).  If you really want to pursue that then it should be a new issue with an explanation why the change makes sense.

My initial thoughts are that having a Loader class is better because it allows the loader to maintain some state bound to that Loader's lifetime (e.g. a cache of loaded files), while having just a global "load" method implies either no state or some hidden global state.